File: Sources/ReminderApp/SpeechCoach.swift
import AVFoundation
import Foundation

/// What the exercise coach needs from a voice: say this now, or be quiet.
/// A protocol so the coach can be driven without a synthesizer in tests and
/// snapshots.
@MainActor
protocol SpeechCoaching: AnyObject {
    /// Interrupts whatever is being said and says `text`. `onFinish` runs
    /// when the whole line has been said — not when it is cut off by `stop`
    /// or by the next `speak`, since whoever cut it off has moved on.
    func speak(_ text: String, onFinish: (() -> Void)?)
    func stop()
}

extension SpeechCoaching {
    func speak(_ text: String) { speak(text, onFinish: nil) }
}

/// The one speech synthesizer in the app, so cues are spoken once however
/// many displays show the takeover, and the Preferences Test button uses the
/// same voice path the coach does.
@MainActor
final class SpeechCoach: NSObject, ObservableObject, SpeechCoaching {
    /// True while an utterance is in progress; the Test button shows it.
    @Published private(set) var isSpeaking = false

    /// The stored preference. Resolved through `VoiceCatalog.resolve` on every
    /// utterance, so a voice the user later uninstalls degrades to the best
    /// available one rather than to silence.
    var voiceIdentifier: String?
    /// `Settings.voiceCoachRate`: percent of normal, 50 = the platform default.
    var rate: Int = 45

    private let synthesizer = AVSpeechSynthesizer()
    /// The completion for the utterance in flight, keyed by the utterance so
    /// a stale delegate callback for an interrupted line cannot fire it.
    private var pending: (utterance: AVSpeechUtterance, onFinish: () -> Void)?

    override init() {
        super.init()
        synthesizer.delegate = self
    }

    func speak(_ text: String, onFinish: (() -> Void)?) {
        speak(text, voiceIdentifier: voiceIdentifier, rate: rate, onFinish: onFinish)
    }

    /// Speaks with a specific voice and pace — the Test button, previewing
    /// the picker's current choice before it is saved.
    func speak(
        _ text: String, voiceIdentifier: String?, rate: Int, onFinish: (() -> Void)? = nil
    ) {
        stop()
        let utterance = AVSpeechUtterance(string: text)
        utterance.voice = VoiceCatalog.resolve(voiceIdentifier)
        utterance.rate = Self.utteranceRate(percent: rate)
        utterance.preUtteranceDelay = 0
        utterance.postUtteranceDelay = 0
        if let onFinish { pending = (utterance, onFinish) }
        synthesizer.speak(utterance)
    }

    func stop() {
        pending = nil
        synthesizer.stopSpeaking(at: .immediate)
        isSpeaking = false
    }

    /// 50 % is `AVSpeechUtteranceDefaultSpeechRate`; the ends of the slider
    /// are clamped well inside AVFoundation's range, where speech is still
    /// speech rather than a drawl or a blur.
    static func utteranceRate(percent: Int) -> Float {
        let clamped = min(max(percent, 20), 80)
        return Float(clamped) / 100
    }

    private func finished(_ utterance: AVSpeechUtterance, completely: Bool) {
        isSpeaking = false
        guard let pending, pending.utterance === utterance else { return }
        self.pending = nil
        if completely { pending.onFinish() }
    }
}

extension SpeechCoach: AVSpeechSynthesizerDelegate {
    nonisolated func speechSynthesizer(
        _ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance
    ) {
        Task { @MainActor in self.isSpeaking = true }
    }

    nonisolated func speechSynthesizer(
        _ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance
    ) {
        Task { @MainActor in self.finished(utterance, completely: true) }
    }

    nonisolated func speechSynthesizer(
        _ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance
    ) {
        Task { @MainActor in self.finished(utterance, completely: false) }
    }
}

/// The system voices worth offering, best first.
enum VoiceCatalog {
    struct Voice: Identifiable, Hashable {
        /// `AVSpeechSynthesisVoice.identifier`, the value stored in settings.
        let id: String
        let name: String
        /// BCP-47, e.g. "en-AU".
        let language: String
        let quality: AVSpeechSynthesisVoiceQuality

        /// "Kate — English (UK) · Enhanced". Some voices already carry the
        /// quality in their name ("Kate (Enhanced)"); it is shown once.
        var label: String {
            let languageName = Locale.current.localizedString(forIdentifier: language) ?? language
            let bareName = name
                .replacingOccurrences(of: " (Enhanced)", with: "")
                .replacingOccurrences(of: " (Premium)", with: "")
            return "\(bareName) — \(languageName) · \(VoiceCatalog.qualityName(quality))"
        }
    }

    /// Installed English voices — the cues are English, so a voice in
    /// another language would mangle them — ordered so the first is the one
    /// to use by default: premium before enhanced before standard, since a
    /// good voice in a neighbouring accent beats a tinny one in your own;
    /// then the user's own region; then by name. Personal voices are left
    /// out (they need their own authorisation) and so are the novelty and
    /// Eloquence voices, which nobody wants coaching them — unless they are
    /// all there is.
    static func installedVoices() -> [Voice] {
        let english = AVSpeechSynthesisVoice.speechVoices().compactMap { voice -> Voice? in
            if #available(macOS 14, *), voice.voiceTraits.contains(.isPersonalVoice) {
                return nil
            }
            guard isEnglish(voice.language) else { return nil }
            return Voice(
                id: voice.identifier, name: voice.name,
                language: voice.language, quality: voice.quality
            )
        }
        let real = english.filter { !isNovelty($0) }
        let candidates = real.isEmpty ? english : real
        return candidates.sorted { lhs, rhs in
            if lhs.quality.rawValue != rhs.quality.rawValue {
                return lhs.quality.rawValue > rhs.quality.rawValue
            }
            let lhsHome = isHomeRegion(lhs), rhsHome = isHomeRegion(rhs)
            if lhsHome != rhsHome { return lhsHome }
            return lhs.name.localizedStandardCompare(rhs.name) == .orderedAscending
        }
    }

    static func defaultVoice() -> Voice? { installedVoices().first }

    /// The voice to speak with: the stored one if it is still installed,
    /// otherwise the best available.
    static func resolve(_ identifier: String?) -> AVSpeechSynthesisVoice? {
        if let identifier, let voice = AVSpeechSynthesisVoice(identifier: identifier) {
            return voice
        }
        return defaultVoice().flatMap { AVSpeechSynthesisVoice(identifier: $0.id) }
    }

    static func qualityName(_ quality: AVSpeechSynthesisVoiceQuality) -> String {
        switch quality {
        case .premium: return "Premium"
        case .enhanced: return "Enhanced"
        default: return "Standard"
        }
    }

    private static func isNovelty(_ voice: Voice) -> Bool {
        voice.id.hasPrefix("com.apple.speech.synthesis.voice.")
            || voice.id.hasPrefix("com.apple.eloquence.")
    }

    private static func isEnglish(_ language: String) -> Bool {
        Locale(identifier: language).language.languageCode?.identifier == "en"
    }

    private static func isHomeRegion(_ voice: Voice) -> Bool {
        guard let region = Locale.current.region?.identifier else { return false }
        return Locale(identifier: voice.language).region?.identifier == region
    }
}

File: Windows/Pauselet.App/SpeechCoach.cs
using System.Speech.Synthesis;
using System.Windows;

namespace Pauselet.App;

/// <summary>
/// What the exercise coach needs from a voice: say this now, or be quiet.
/// An interface so the coach can be driven without a synthesizer in tests.
/// (Mirrors SpeechCoaching in SpeechCoach.swift.)
/// </summary>
internal interface ISpeechCoaching
{
    /// <summary>
    /// Interrupts whatever is being said and says <paramref name="text"/>.
    /// <paramref name="onFinish"/> runs when the whole line has been said —
    /// not when it is cut off by <see cref="Stop"/> or by the next
    /// <see cref="Speak"/>, since whoever cut it off has moved on.
    /// </summary>
    void Speak(string text, Action? onFinish = null);

    void Stop();
}

/// <summary>
/// The one speech synthesizer in the app, so cues are spoken once however
/// many displays show the takeover, and the Preferences Test button uses the
/// same voice path the coach does.
///
/// System.Speech rather than the WinRT synthesizer: it speaks straight to the
/// default output device with no media-player plumbing, which is all a spoken
/// cue needs, and it enumerates the same SAPI voices the Windows speech
/// settings install.
/// </summary>
internal sealed class SpeechCoach : ISpeechCoaching, IDisposable
{
    /// <summary>True while an utterance is in progress; the Test button shows it.</summary>
    public bool IsSpeaking { get; private set; }

    /// <summary>Raised on the UI thread when <see cref="IsSpeaking"/> changes.</summary>
    public event Action? SpeakingChanged;

    /// <summary>
    /// The stored preference. Resolved through <see cref="VoiceCatalog"/> on
    /// every utterance, so a voice the user later uninstalls degrades to the
    /// best available one rather than to silence.
    /// </summary>
    public string? VoiceIdentifier { get; set; }

    /// <summary>Settings.VoiceCoachRate: percent of normal, 50 = the platform default.</summary>
    public int Rate { get; set; } = 45;

    private readonly SpeechSynthesizer _synthesizer = new();
    /// <summary>
    /// The completion for the utterance in flight, keyed by the prompt so a
    /// stale completion for an interrupted line cannot fire it.
    /// </summary>
    private Prompt? _pendingPrompt;
    private Action? _pendingFinish;
    private bool _disposed;

    public SpeechCoach()
    {
        _synthesizer.SetOutputToDefaultAudioDevice();
        _synthesizer.SpeakStarted += (_, _) => SetSpeaking(true);
        _synthesizer.SpeakCompleted += OnSpeakCompleted;
    }

    public void Speak(string text, Action? onFinish = null) =>
        Speak(text, VoiceIdentifier, Rate, onFinish);

    /// <summary>
    /// Speaks with a specific voice and pace — the Test button, previewing the
    /// picker's current choice before it is saved.
    /// </summary>
    public void Speak(string text, string? voiceIdentifier, int rate, Action? onFinish = null)
    {
        if (_disposed) return;
        Stop();
        if (VoiceCatalog.Resolve(voiceIdentifier) is { } voiceName)
        {
            try
            {
                _synthesizer.SelectVoice(voiceName);
            }
            catch (ArgumentException)
            {
                // The voice went away between enumeration and now; the
                // synthesizer keeps whichever one it already had.
            }
        }
        _synthesizer.Rate = UtteranceRate(rate);
        var prompt = new Prompt(text);
        _pendingPrompt = prompt;
        _pendingFinish = onFinish;
        _synthesizer.SpeakAsync(prompt);
    }

    public void Stop()
    {
        if (_disposed) return;
        _pendingPrompt = null;
        _pendingFinish = null;
        _synthesizer.SpeakAsyncCancelAll();
        SetSpeaking(false);
    }

    /// <summary>
    /// The settings slider is percent of normal, 50 being the default;
    /// System.Speech's Rate is -10…10 around its own default of 0. The ends
    /// are clamped well inside the range, where speech is still speech rather
    /// than a drawl or a blur.
    /// </summary>
    public static int UtteranceRate(int percent)
    {
        var clamped = Math.Clamp(percent, 20, 80);
        // 20 % → -6, 50 % → 0, 80 % → +6.
        return (int)Math.Round((clamped - 50) / 5.0);
    }

    private void OnSpeakCompleted(object? sender, SpeakCompletedEventArgs e)
    {
        // System.Speech raises this on a worker thread, and the completion
        // releases the coach's announcement — which walks the session and the
        // overlay. Both belong to the UI thread, so the whole handler is
        // marshalled rather than just the flag; that also keeps the pending
        // prompt's bookkeeping single-threaded.
        var dispatcher = Application.Current?.Dispatcher;
        if (dispatcher is not null && !dispatcher.CheckAccess())
        {
            dispatcher.BeginInvoke(() => OnSpeakCompleted(sender, e));
            return;
        }

        SetSpeaking(false);
        // A stale completion for a prompt that has since been superseded must
        // not release the announcement of the phase that replaced it.
        if (_pendingPrompt is null || !ReferenceEquals(_pendingPrompt, e.Prompt)) return;
        var finish = _pendingFinish;
        _pendingPrompt = null;
        _pendingFinish = null;
        // Cancelled prompts also complete; only a line that was actually said
        // through releases the coach's announcement.
        if (!e.Cancelled) finish?.Invoke();
    }

    /// <summary>
    /// Synthesizer events arrive on a worker thread; the coach and the Test
    /// button both touch WPF state, so hop to the UI thread.
    /// </summary>
    private void SetSpeaking(bool value)
    {
        var dispatcher = Application.Current?.Dispatcher;
        if (dispatcher is not null && !dispatcher.CheckAccess())
        {
            dispatcher.BeginInvoke(() => SetSpeaking(value));
            return;
        }
        if (IsSpeaking == value) return;
        IsSpeaking = value;
        SpeakingChanged?.Invoke();
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _synthesizer.SpeakAsyncCancelAll();
        _synthesizer.Dispose();
    }
}

/// <summary>
/// The system voices worth offering, best first. (Mirrors VoiceCatalog in
/// SpeechCoach.swift; SAPI has no quality tier, so the ordering is by region
/// then name.)
/// </summary>
internal static class VoiceCatalog
{
    /// <param name="Id">The SAPI voice name, the value stored in settings.</param>
    /// <param name="Name">The voice's own name, e.g. "Microsoft Zira Desktop".</param>
    /// <param name="Language">BCP-47, e.g. "en-AU".</param>
    internal sealed record Voice(string Id, string Name, string Language)
    {
        /// <summary>"Zira — English (United States)".</summary>
        public string Label
        {
            get
            {
                string languageName;
                try
                {
                    languageName = new System.Globalization.CultureInfo(Language).DisplayName;
                }
                catch (System.Globalization.CultureNotFoundException)
                {
                    languageName = Language;
                }
                // SAPI names are "Microsoft Zira Desktop"; the middle word is
                // the only part that identifies the voice to a person.
                var bare = Name
                    .Replace("Microsoft ", "")
                    .Replace(" Desktop", "");
                return $"{bare} — {languageName}";
            }
        }
    }

    /// <summary>
    /// Enumerating voices constructs a synthesizer and walks the SAPI
    /// registry, which is far too slow to do per spoken cue. The list only
    /// changes when the user installs a voice in Windows' settings, so it is
    /// read once per app run.
    /// </summary>
    private static IReadOnlyList<Voice>? _cached;

    /// <summary>
    /// Installed English voices — the cues are English, so a voice in another
    /// language would mangle them — ordered so the first is the one to use by
    /// default: the user's own region first, then by name.
    /// </summary>
    public static IReadOnlyList<Voice> InstalledVoices() => _cached ??= Enumerate();

    private static IReadOnlyList<Voice> Enumerate()
    {
        using var synthesizer = new SpeechSynthesizer();
        List<Voice> english;
        try
        {
            english = synthesizer.GetInstalledVoices()
                .Where(voice => voice.Enabled)
                .Select(voice => voice.VoiceInfo)
                .Where(info => IsEnglish(info.Culture.Name))
                .Select(info => new Voice(info.Name, info.Name, info.Culture.Name))
                .ToList();
        }
        catch (PlatformNotSupportedException)
        {
            return [];
        }
        return english
            .OrderByDescending(IsHomeRegion)
            .ThenBy(voice => voice.Name, StringComparer.CurrentCultureIgnoreCase)
            .ToList();
    }

    public static Voice? DefaultVoice() => InstalledVoices().FirstOrDefault();

    /// <summary>
    /// The voice to speak with: the stored one if it is still installed,
    /// otherwise the best available. <c>null</c> when there are no English
    /// voices at all, which leaves the synthesizer on its own default.
    /// </summary>
    public static string? Resolve(string? identifier)
    {
        var installed = InstalledVoices();
        if (identifier is { Length: > 0 }
            && installed.Any(voice => voice.Id == identifier))
        {
            return identifier;
        }
        return installed.FirstOrDefault()?.Id;
    }

    private static bool IsEnglish(string culture) =>
        culture.StartsWith("en", StringComparison.OrdinalIgnoreCase);

    private static bool IsHomeRegion(Voice voice)
    {
        var region = System.Globalization.RegionInfo.CurrentRegion.TwoLetterISORegionName;
        var parts = voice.Language.Split('-');
        return parts.Length > 1
            && string.Equals(parts[^1], region, StringComparison.OrdinalIgnoreCase);
    }
}
